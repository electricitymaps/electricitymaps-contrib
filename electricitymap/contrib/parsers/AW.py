#!/usr/bin/env python3

from datetime import datetime
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

FUEL_MAPPING = {
    "Fossil": "oil",
    "Wind": "wind",
    "TotalSolar": "solar",
    "total_bio_gas": "biomass",
}
TIMEZONE = ZoneInfo("America/Aruba")
PRODUCTION_URL = (
    "https://www.webaruba.com/renewable-energy-dashboard/app/rest/results.json"
)
SOURCE = "webaruba.com"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AW"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict[str, Any]]:
    if target_datetime:
        raise NotImplementedError("This parser is not yet able to parse past dates")

    r = session or Session()
    # User agent is mandatory or services answers 404
    headers = {
        "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko)"
    }
    response = r.get(PRODUCTION_URL, headers=headers)
    aruba_json = response.json()
    top_data = aruba_json["dashboard_top_data"]
    mix = ProductionMix()
    sources_total = 0
    for fuel, mode in FUEL_MAPPING.items():
        value = float(top_data[fuel]["value"])
        mix.add_value(mode, value)
        sources_total += value
    # "unknown" is when data reported in the categories above is less than total reported.
    # If categories sum up to more than total, accept the datapoint, but only if it's less than 10% of total.
    # This helps avoid missing data when it's a little bit off, due to rounding or reporting
    reported_total = float(top_data["TotalPower"]["value"])

    if (sources_total / reported_total) > 1.1:
        raise ParserException(
            "AW.py",
            f"AW parser reports fuel sources add up to {sources_total} but total generation {reported_total} is lower",
            zone_key,
        )

    mix.add_value(
        "unknown", reported_total - sources_total, correct_negative_with_zero=True
    )
    # We're using Fossil data to get timestamp in correct time zone
    local_date_time = datetime.strptime(
        top_data["total_bio_gas"]["timestamp"], "%Y-%m-%d %H:%M:%S.%f"
    ).replace(tzinfo=TIMEZONE)

    production_list = ProductionBreakdownList(logger)
    production_list.append(
        zone_key,
        local_date_time,
        SOURCE,
        production=mix,
    )

    return production_list.to_list()


if __name__ == "__main__":
    print(fetch_production())
