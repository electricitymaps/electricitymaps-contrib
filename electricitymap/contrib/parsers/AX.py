from datetime import datetime, timedelta
from logging import Logger, getLogger
from re import findall
from zoneinfo import ZoneInfo

from bs4 import BeautifulSoup
from requests import Response, Session

from electricitymap.contrib.lib.models.event_lists import (
    ExchangeList,
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey

from .lib.exceptions import ParserException

IFRAME_URL = "https://grafik.kraftnat.ax/grafer/tot_inm_24h_15.php"
TIME_ZONE = ZoneInfo("Europe/Mariehamn")
SOURCE = "kraftnat.ax"


def fetch_data(session: Session, logger: Logger):
    """Fetch data from the iFrame."""

    response: Response = session.get(IFRAME_URL)
    response.raise_for_status()

    soup = BeautifulSoup(response.text, "html.parser").find_all("script")
    result_time_series = findall(r"data: \[(.*?)\]", str(soup))
    if len(result_time_series) != 3:
        raise ParserException(
            "AX.py",
            "Did not find the expected amount of results. Check if the website has changed.",
        )
    time_series: list = result_time_series[0].split(",")
    raw_data: list[str] = findall(r"data:\[(.*?)\]", str(soup))
    if len(raw_data) != 6:
        raise ParserException(
            "AX.py",
            "The raw data did not match the expected format. Check if the website has changed.",
        )
    for raw in raw_data:
        if len(raw.split(",")) != len(time_series):
            raise ParserException(
                "AX.py",
                "The raw data did not match the length of the the time series. Check if the website has changed.",
            )
    data_list = []
    for time, sweden, alink, fossil, gustavs, wind, consumption in zip(
        time_series,
        raw_data[0].split(","),
        raw_data[1].split(","),
        raw_data[2].split(","),
        raw_data[3].split(","),
        raw_data[4].split(","),
        raw_data[5].split(","),
        strict=True,
    ):
        data_list.append(
            {
                "time": str(time.replace('"', "")),
                "sweden": float(sweden),
                "alink": float(alink),
                "fossil": float(fossil),
                "gustavs": float(gustavs),
                "wind": float(wind),
                "consumption": float(consumption),
            }
        )
    return data_list


def formatted_data(
    zone_key: ZoneKey | None,
    zone_key1: ZoneKey | None,
    zone_key2: ZoneKey | None,
    session: Session,
    logger: Logger,
    data_type: str,
):
    """Format data to Electricity Map standards."""
    data_list = fetch_data(session, logger)
    data_list.reverse()
    date_time = datetime.now(TIME_ZONE)
    date = date_time.replace(
        hour=int(data_list[0]["time"].split(":")[0]),
        minute=int(data_list[0]["time"].split(":")[1]),
        second=0,
        microsecond=0,
    )
    if date > date_time:
        date = date - timedelta(days=1)
    exchanges = ExchangeList(logger)
    production_breakdowns = ProductionBreakdownList(logger)
    consumption = TotalConsumptionList(logger)
    for data in data_list:
        corrected_date = date - timedelta(minutes=15 * data_list.index(data))
        if data_type == "production" and zone_key is not None:
            production_mix = ProductionMix(wind=data["wind"], oil=data["fossil"])
            production_breakdowns.append(
                datetime=corrected_date,
                production=production_mix,
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "consumption" and zone_key is not None:
            consumption.append(
                datetime=corrected_date,
                consumption=data["consumption"],
                source=SOURCE,
                zoneKey=zone_key,
            )
        elif data_type == "exchange":
            if zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("SE-SE3"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=data["sweden"] * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->SE-SE3"),
                )
            elif zone_key1 == ZoneKey("AX") and zone_key2 == ZoneKey("FI"):
                exchanges.append(
                    datetime=corrected_date,
                    netFlow=round(data["alink"] + data["gustavs"], 2) * -1,
                    source=SOURCE,
                    zoneKey=ZoneKey("AX->FI"),
                )
            else:
                raise ParserException(
                    "AX.py",
                    "This parser can only fetch data between Åland <-> Sweden and Åland <-> Finland",
                )
        else:
            raise ParserException(
                "AX.py",
                "The datasource currently implemented is only for production, consumption and exchange data",
                zone_key,
            )

    if data_type == "production":
        return production_breakdowns.to_list()
    elif data_type == "consumption":
        return consumption.to_list()
    elif data_type == "exchange":
        return exchanges.to_list()
    else:
        raise ParserException(
            "AX.py",
            "The datasource currently implemented is only for production, consumption and exchange data",
            zone_key,
        )


def fetch_production(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch production data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="production",
    )


def fetch_consumption(
    zone_key: ZoneKey = ZoneKey("AX"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch consumption data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=zone_key,
        zone_key1=None,
        zone_key2=None,
        session=session,
        logger=logger,
        data_type="consumption",
    )


def fetch_exchange(
    zone_key1: ZoneKey,
    zone_key2: ZoneKey,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Fetch exchange data."""

    if target_datetime is not None:
        raise ParserException(
            "AX.py", "The datasource currently implemented is only for real time data"
        )

    return formatted_data(
        zone_key=None,
        zone_key1=zone_key1,
        zone_key2=zone_key2,
        session=session,
        logger=logger,
        data_type="exchange",
    )
