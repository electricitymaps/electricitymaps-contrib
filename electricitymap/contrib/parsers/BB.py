from datetime import datetime, timedelta
from logging import Logger, getLogger
from typing import Any
from zoneinfo import ZoneInfo

import pandas as pd

# The request library is used to fetch content through HTTP
import requests
from requests import Session
from requests.auth import AuthBase

from electricitymap.contrib.lib.models.event_lists import (
    ProductionBreakdownList,
    TotalConsumptionList,
)
from electricitymap.contrib.lib.models.events import ProductionMix, StorageMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.config import refetch_frequency
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "energy.gov.bb"
base_url = "https://dataservices.acelerex.com/api"


class JwtPublicAuth(AuthBase):
    def __init__(self):
        self.token = None

    def __call__(self, r):
        if self.token is None:
            self.token = self.get_token()
        r.headers["Authorization"] = f"Bearer {self.token}"
        return r

    def get_token(self) -> str:
        url = base_url + "/authentication"
        response = requests.post(
            url, json={"dashboardId": "barbados", "strategy": "jwt-public"}
        )
        return response.json()["accessToken"]


def session_with_auth(session: Session) -> Session:
    if session.auth is None:
        session.auth = JwtPublicAuth()
    return session


def parse_operational_data(response_json) -> tuple[list[str], list[list[str]]]:
    header, *table = (x for x in response_json["data"][0]["data"]["data"] if len(x) > 1)
    return header, table


def fetch_operational_data(
    zone_key,
    session: Session,
    target_datetime: datetime | None,
    logger: Logger,
) -> pd.DataFrame:
    session = session_with_auth(session)

    params = {"name": "operational-data", "$sort[created_at]": -1}

    header = None
    table = None

    if target_datetime is None:
        data = session.get(base_url + "/barbados/uploads", params=params).json()
        header, table = parse_operational_data(data)
    elif target_datetime > datetime(year=2023, month=2, day=27):
        expected_date = target_datetime.astimezone(
            ZoneInfo("America/Barbados")
        ).strftime("%Y-%m-%d")

        for i in range(10):
            url_date = (
                (target_datetime + timedelta(days=i))
                .astimezone(ZoneInfo("America/Barbados"))
                .strftime("%Y-%m-%d")
            )
            params["created_at[$lt]"] = url_date
            data = session.get(base_url + "/barbados/uploads", params=params).json()
            header, table = parse_operational_data(data)
            if table[0][0][:10] == expected_date:
                break
            if table[0][0][:10] > expected_date:
                raise ParserException("BB.py", "Date not found", zone_key)
    else:
        raise ParserException(
            "BB.py",
            "This parser is not yet able to parse dates before 2023-02-27",
            zone_key,
        )

    if table is None or header is None:
        raise ParserException("BB.py", "No data found", zone_key)

    operational_data = pd.DataFrame(
        table,
        columns=header,
    )
    columns = [
        "24 hr monitored solar production",
        "24 hr monitored wind production",
        "24 hr storage charging",
        "24 hr storage discharging",
        "24 hr fossil production",
        "24 hr Net demand",
        "24 hr Curtailment",
    ]
    for columm in columns:
        operational_data[columm] = pd.to_numeric(operational_data[columm])
    return operational_data


@refetch_frequency(timedelta(days=1))
def fetch_production(
    zone_key: ZoneKey = ZoneKey("BB"),
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict] | dict:
    operational_data = fetch_operational_data(
        zone_key, session, target_datetime, logger
    )

    production_list = ProductionBreakdownList(logger=logger)

    for _index, row in operational_data.iterrows():
        production_list.append(
            zoneKey=zone_key,
            datetime=datetime.fromisoformat(row["datetime"]),
            production=ProductionMix(
                solar=row["24 hr monitored solar production"],
                wind=row["24 hr monitored wind production"],
                oil=row["24 hr fossil production"],
            ),
            storage=StorageMix(
                battery=(
                    row["24 hr storage discharging"] + row["24 hr storage charging"]
                )
                * -1
            ),
            source=SOURCE,
        )
    return production_list.to_list()


@refetch_frequency(timedelta(days=1))
def fetch_consumption(
    zone_key: ZoneKey,
    session: Session = Session(),
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> dict[str, Any] | list[dict[str, Any]]:
    operational_data = fetch_operational_data(
        zone_key, session, target_datetime, logger
    )

    consumption_list = TotalConsumptionList(logger=logger)

    for _index, row in operational_data.iterrows():
        consumption_list.append(
            zoneKey=zone_key,
            datetime=datetime.fromisoformat(row["datetime"]),
            consumption=row["24 hr Net demand"],
            source=SOURCE,
        )
    return consumption_list.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Maps backend, but handy for testing."""

    print("fetch_production(XX) ->")
    print(fetch_production(ZoneKey("BB")))
    fetch_production(ZoneKey("BB"), target_datetime=datetime(2024, 2, 2))
    print("fetch_consumption(XX) ->")
    print(fetch_consumption(ZoneKey("BB")))
