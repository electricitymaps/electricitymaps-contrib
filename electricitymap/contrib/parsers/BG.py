from datetime import datetime, timezone
from logging import Logger, getLogger

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

PARSER = "BG.py"

SOURCE = "eso.bg"
SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php?en"

PRODUCTION_TYPE_TO_PRODUCTION_MODE = {
    "NPP": "nuclear",
    "CHP": "coal",
    "Heating TPPs": "gas",
    "Factory TPPs": "gas",
    "HPP": "hydro",
    "Small HPPs": "hydro",
    "Wind": "wind",
    "PV": "solar",
    "Bio": "biomass",
}


def fetch_production(
    zone_key: ZoneKey = ZoneKey("BG"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    session = session or Session()

    if target_datetime is not None:
        raise ParserException(
            PARSER, "This parser is not yet able to parse historical data", zone_key
        )

    time = datetime.now(timezone.utc)
    response = session.get(SOURCE_API_URL)
    if not response.ok:
        raise ParserException(
            PARSER,
            f"Exception when fetching production error code: {response.status_code}: {response.text}",
            zone_key,
        )

    response_payload = response.json()
    logger.debug(f"Raw generation breakdown: {response_payload}")

    production_mix = ProductionMix()
    for header, value in response_payload:
        production_type, _delimiter, _percentage = header.rpartition(" ")
        production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE[production_type]
        production_mix.add_value(production_mode, value)

    production_breakdown_list = ProductionBreakdownList(logger)
    production_breakdown_list.append(
        zoneKey=zone_key,
        datetime=time,
        source=SOURCE,
        production=production_mix,
    )
    return production_breakdown_list.to_list()
