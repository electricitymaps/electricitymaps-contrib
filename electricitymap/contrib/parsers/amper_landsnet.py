from datetime import datetime
from logging import Logger, getLogger
from zoneinfo import ZoneInfo

from requests import Session

from electricitymap.contrib.lib.models.event_lists import ProductionBreakdownList
from electricitymap.contrib.lib.models.events import ProductionMix
from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers.lib.exceptions import ParserException

SOURCE = "amper.landsnet.is"
SOURCE_URL = "https://amper.landsnet.is/generation/api/Values"


def fetch_production(
    zone_key: ZoneKey = ZoneKey("IS"),
    session: Session | None = None,
    target_datetime: datetime | None = None,
    logger: Logger = getLogger(__name__),
) -> list[dict]:
    """Requests the last known production mix (in MW) of a given country."""
    r = session or Session()
    if target_datetime is not None:
        raise NotImplementedError("This parser is not yet able to parse past dates")
    res = r.get(SOURCE_URL)
    if not res.ok:
        raise ParserException(
            parser="amper_landsnet.py",
            message=f"Failed to fetch {SOURCE}: {res.status_code}, err: {res.text}",
            zone_key=zone_key,
        )

    obj = res.json()
    mix = ProductionMix(
        geothermal=obj["geothermal"],
        hydro=obj["hydro"],
        oil=obj["oil"],
    )
    breakdowns = ProductionBreakdownList(logger=logger)
    breakdowns.append(
        zoneKey=zone_key,
        datetime=datetime.fromisoformat(obj["timestamp"]).replace(
            tzinfo=ZoneInfo("Atlantic/Reykjavik")
        ),
        production=mix,
        source=SOURCE,
    )
    return breakdowns.to_list()


if __name__ == "__main__":
    """Main method, never used by the Electricity Map backend, but handy for testing."""

    print("fetch_production() ->")
    print(fetch_production())
