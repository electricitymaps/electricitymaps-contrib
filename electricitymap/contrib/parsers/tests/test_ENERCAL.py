from importlib import resources

from requests_mock import ANY, GET

from electricitymap.contrib.lib.types import ZoneKey
from electricitymap.contrib.parsers import ENERCAL


def test_production_with_snapshot(adapter, session, snapshot):
    raw_data = (
        resources.files("electricitymap.contrib.parsers.tests.mocks.ENERCAL")
        .joinpath("production.json")
        .read_bytes()
    )
    adapter.register_uri(
        GET,
        ANY,
        content=raw_data,
    )
    assert snapshot == ENERCAL.fetch_production(ZoneKey("NC"), session)
