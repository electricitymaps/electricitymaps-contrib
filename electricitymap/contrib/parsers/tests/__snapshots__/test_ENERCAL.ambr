# serializer version: 1
# name: test_production_with_snapshot
  list([
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 0, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 23.7,
        'hydro': 35.9,
        'oil': 32,
        'solar': 0.3,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 1, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 20.7,
        'hydro': 35.2,
        'oil': 32.3,
        'solar': 0,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 2, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 16.8,
        'hydro': 35.7,
        'oil': 33.8,
        'solar': 0,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 3, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 21.1,
        'hydro': 28.8,
        'oil': 32,
        'solar': 0,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 4, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 28.4,
        'hydro': 21,
        'oil': 32.6,
        'solar': 0,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 5, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 27.8,
        'hydro': 23.5,
        'oil': 33.6,
        'solar': 0.1,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 6, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 16.6,
        'hydro': 34.5,
        'oil': 33.7,
        'solar': 3.6,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 7, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 20.2,
        'hydro': 14,
        'oil': 32.5,
        'solar': 20,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 8, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 18.3,
        'hydro': 0.3,
        'oil': 27.3,
        'solar': 34.6,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 9, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 12.8,
        'hydro': 0.8,
        'oil': 9.5,
        'solar': 55.3,
        'wind': 0.2,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 10, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 11.5,
        'hydro': 1.4,
        'oil': 5.9,
        'solar': 53.1,
        'wind': 1.3,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 11, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 9.4,
        'hydro': 6.8,
        'oil': 10.9,
        'solar': 48.9,
        'wind': 2,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 12, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 13.1,
        'hydro': 10.2,
        'oil': 27.9,
        'solar': 41.5,
        'wind': 3.4,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 13, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 16.5,
        'hydro': 1.4,
        'oil': 23.5,
        'solar': 39.2,
        'wind': 3.2,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 14, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 21.5,
        'hydro': 6.8,
        'oil': 17.8,
        'solar': 24.6,
        'wind': 2.4,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 15, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 18.5,
        'hydro': 17.2,
        'oil': 32,
        'solar': 18.7,
        'wind': 2.1,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 16, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 18.4,
        'hydro': 34.6,
        'oil': 36.4,
        'solar': 14.2,
        'wind': 2.2,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 17, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 20,
        'hydro': 34.6,
        'oil': 33.9,
        'solar': 10.3,
        'wind': 0.6,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 18, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 19.8,
        'hydro': 37.5,
        'oil': 34.9,
        'solar': 3.1,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 19, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 21.3,
        'hydro': 38.7,
        'oil': 42.4,
        'solar': 1.2,
        'wind': 0,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 20, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 21.5,
        'hydro': 39,
        'oil': 44,
        'solar': 0.5,
        'wind': 0.1,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 21, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 20.2,
        'hydro': 38.7,
        'oil': 42.5,
        'solar': 0,
        'wind': 1.9,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 22, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 19,
        'hydro': 38.6,
        'oil': 37.1,
        'solar': 0,
        'wind': 3.2,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
    dict({
      'correctedModes': list([
      ]),
      'datetime': datetime.datetime(2024, 1, 1, 23, 0, tzinfo=zoneinfo.ZoneInfo(key='Pacific/Noumea')),
      'production': dict({
        'coal': 17.3,
        'hydro': 38.4,
        'oil': 32,
        'solar': 0.3,
        'wind': 2.8,
      }),
      'source': 'enercal.nc',
      'sourceType': <EventSourceType.measured: 'measured'>,
      'storage': dict({
      }),
      'zoneKey': 'NC',
    }),
  ])
# ---
